File: library/xml-model/src/main/java/org/cru/godtools/xml/model/DeviceType.java
package org.cru.godtools.xml.model;

import com.annimon.stream.Stream;
import com.google.common.collect.Sets;

import org.jetbrains.annotations.Contract;

import java.util.EnumSet;
import java.util.Set;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.VisibleForTesting;

enum DeviceType {
    MOBILE, UNKNOWN;

    public static final Set<DeviceType> ALL = Sets.immutableEnumSet(EnumSet.allOf(DeviceType.class));

    @VisibleForTesting
    static final String XML_DEVICE_TYPE_MOBILE = "mobile";

    @Nullable
    @Contract("!null -> !null; null -> null")
    static DeviceType parseSingle(@Nullable final String type) {
        if (type == null) {
            return null;
        }

        switch (type) {
            case XML_DEVICE_TYPE_MOBILE:
                return MOBILE;
            default:
                return UNKNOWN;
        }

    }

    @Nullable
    @Contract("_,!null -> !null; !null,_ -> !null")
    static Set<DeviceType> parse(@Nullable final String types, @NonNull final Set<DeviceType> defValue) {
        if (types == null) {
            return defValue;
        }

        return Sets.immutableEnumSet(
                Stream.of(types.split("\\s+"))
                        .map(DeviceType::parseSingle)
                        .distinct()
                        .toList()
        );
    }
}

File: library/xml-model/src/main/java/org/cru/godtools/xml/model/DeviceType.kt
package org.cru.godtools.xml.model

import androidx.annotation.VisibleForTesting
import java.util.Collections
import java.util.EnumSet

enum class DeviceType {
    ANDROID, MOBILE, UNKNOWN;

    companion object {
        internal val ALL: Set<DeviceType> = Collections.unmodifiableSet(EnumSet.allOf(DeviceType::class.java))

        @VisibleForTesting
        internal const val XML_DEVICE_TYPE_ANDROID = "android"
        @VisibleForTesting
        internal const val XML_DEVICE_TYPE_MOBILE = "mobile"

        @VisibleForTesting
        internal fun parseSingle(type: String?) = when (type) {
            null -> null
            XML_DEVICE_TYPE_ANDROID -> ANDROID
            XML_DEVICE_TYPE_MOBILE -> MOBILE
            else -> UNKNOWN
        }

        internal fun parse(types: String?, defValue: Set<DeviceType>) = when {
            types != null -> REGEX_SEQUENCE_SEPARATOR.split(types).mapNotNullTo(mutableSetOf()) { parseSingle(it) }
            else -> defValue
        }
    }
}
